File: src/jsonl-reader.ts
import fs from "node:fs";
import path from "node:path";
import { encodeProjectPath, getProjectsDir } from "./path-utils.js";

export function extractCwdFromJsonl(jsonlPath: string): string | undefined {
  let fd: number | undefined;
  try {
    fd = fs.openSync(jsonlPath, "r");
    const buf = Buffer.alloc(8192);
    const bytesRead = fs.readSync(fd, buf, 0, 8192, 0);

    const chunk = buf.toString("utf-8", 0, bytesRead);
    const lines = chunk.split("\n");
    for (let i = 0; i < Math.min(lines.length, 20); i++) {
      try {
        const record = JSON.parse(lines[i]);
        if (typeof record.cwd === "string") return record.cwd;
      } catch { continue; }
    }
  } catch { /* ignore */ }
  finally {
    if (fd !== undefined) fs.closeSync(fd);
  }
  return undefined;
}

function sessionJsonlPath(sessionId: string, cwd: string, claudeHome = "~/.claude"): string | undefined {
  const encoded = encodeProjectPath(cwd);
  const file = path.join(getProjectsDir(claudeHome), encoded, `${sessionId}.jsonl`);
  return fs.existsSync(file) ? file : undefined;
}

function extractUserText(content: unknown): string {
  if (typeof content === "string") return content;
  if (Array.isArray(content)) {
    return content
      .filter((b): b is Record<string, unknown> => typeof b === "object" && b !== null && b.type === "text")
      .map((b) => (typeof b.text === "string" ? b.text : ""))
      .join(" ");
  }
  return "";
}

export function getLastUserMessages(
  sessionId: string,
  cwd: string,
  n = 3,
  maxChars = 200,
  claudeHome = "~/.claude",
): string[] {
  const filePath = sessionJsonlPath(sessionId, cwd, claudeHome);
  if (!filePath) return [];

  const messages: string[] = [];
  try {
    const lines = fs.readFileSync(filePath, "utf-8").split("\n");
    for (const line of lines) {
      let record: Record<string, unknown>;
      try {
        record = JSON.parse(line);
      } catch {
        continue;
      }
      if (record.type !== "user") continue;
      const msg = record.message as Record<string, unknown> | undefined;
      if (!msg) continue;
      const text = extractUserText(msg.content);
      if (text) messages.push(text.slice(0, maxChars));
    }
  } catch {
    return [];
  }

  return messages.slice(-n);
}

export function extractSummary(
  sessionId: string,
  cwd: string,
  maxMessages = 5,
  maxChars = 200,
  claudeHome = "~/.claude",
): string {
  const messages = getLastUserMessages(sessionId, cwd, maxMessages, maxChars, claudeHome);
  if (messages.length === 0) return "No messages in session";
  return messages.join(" | ");
}

export function getMessageCount(sessionId: string, cwd: string, claudeHome = "~/.claude"): number {
  return getLastUserMessages(sessionId, cwd, 9999, 1, claudeHome).length;
}

File: src/cc_proxy/jsonl_reader.py
"""Parse Claude Code JSONL session history for summaries."""

import json
import os
from pathlib import Path

from .path_utils import encode_project_path, get_claude_projects_dir


def _session_jsonl_path(session_id: str, cwd: str, claude_home: str = "~/.claude") -> Path | None:
    encoded = encode_project_path(cwd)
    jsonl_file = Path(get_claude_projects_dir(claude_home)) / encoded / f"{session_id}.jsonl"
    return jsonl_file if jsonl_file.exists() else None


def _extract_user_text(content) -> str:
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        parts = []
        for block in content:
            if isinstance(block, dict) and block.get("type") == "text":
                parts.append(block.get("text", ""))
        return " ".join(parts)
    return ""


def get_last_user_messages(
    session_id: str,
    cwd: str,
    n: int = 3,
    max_chars: int = 200,
    claude_home: str = "~/.claude",
) -> list[str]:
    jsonl_path = _session_jsonl_path(session_id, cwd, claude_home)
    if jsonl_path is None:
        return []

    user_messages: list[str] = []
    try:
        with open(jsonl_path) as f:
            for line in f:
                try:
                    record = json.loads(line.strip())
                except json.JSONDecodeError:
                    continue

                if record.get("type") != "user":
                    continue
                msg = record.get("message", {})
                text = _extract_user_text(msg.get("content"))
                if text:
                    user_messages.append(text[:max_chars])
    except OSError:
        return []

    return user_messages[-n:]


def extract_summary(
    session_id: str,
    cwd: str,
    max_messages: int = 5,
    max_chars: int = 200,
    claude_home: str = "~/.claude",
) -> str:
    messages = get_last_user_messages(session_id, cwd, max_messages, max_chars, claude_home)
    if not messages:
        return "No messages in session"
    return " | ".join(messages)


def get_message_count(session_id: str, cwd: str, claude_home: str = "~/.claude") -> int:
    return len(get_last_user_messages(session_id, cwd, n=9999, max_chars=1, claude_home=claude_home))
